### PR TITLE
Fix: added async in allow method in no-empty-function (fixes #12768)

### LINKED ISSUE #12768: no-empty-function false positive
**Tell us about your environment**

* **ESLint Version:** 6.8.0
* **Node Version:** 12.14.1
* **npm Version:** 6.13.4

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```json
{
  "no-empty-function": ["error", { "allow": ["methods"] }]
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const object = {
  async method() {}
};
```

```bash
eslint .
```

**What did you expect to happen?**

No `no-empty-function` warnings.

**What actually happened? Please include the actual, raw output from ESLint.**

> Unexpected empty async method 'method'. (no-empty-function)

**Are you willing to submit a pull request to fix this bug?**

Yes

### PR DESCRIPTION
**Pull request by anikethsaha**
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).
- [X] The team has reached consensus on the changes proposed in this pull request. If not, I understand that the evaluation process will begin with this pull request and won't be merged until the team has reached consensus.

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[X] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[X] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Added `asyncFunctions` and `asyncMethods` to `ALLOW_OPTIONS` for `no-empty-function` rule

#### Is there anything you'd like reviewers to focus on?

fixes https://github.com/eslint/eslint/issues/12768

Also, I am bit confused about how to write tests for valid code as for this rule, gives test cases are converted into valid and invalid

### CODE REVIEW

**Comment by anikethsaha** [comment]
> The test setup is a bit confusing with use of Array.prototype.reduce(). If it's helpful to refactor this to make it easier to understand what's going on, I think that would be fine.

indeed this is a bit confusing. 
I tried adding async code in the `initialValue` for the `reduce` method  along with the current one
[`here`](https://github.com/eslint/eslint/blob/master/tests/lib/rules/no-empty-function.js#L266-L271)


I added this

image


 but got parsing error

image

**Comment by yeonjuan** [comment]
@anikethsaha Thanks for working on it 👍 

> but got parsing error

You should change the  parserOptions  with `ecmaVersion >= 8` ( 8, 9, ..)

**Comment by anikethsaha** [comment]
thanks a lot @yeonjuan ! it worked like a charm. 
_(I thought async are supported in es2015 !)_

@kaicataldo I added two test cases. Just do let me know if you need to add some more!

File: tests/lib/rules/no-empty-function.js [hunk 1]
```diff
@@ -267,6 +269,16 @@ ruleTester.run("no-empty-function", rule, [
...
+        },
+        {
+            code: "const foo = { async method() {} }",
+            options: [{ allow: ["asyncMethods"] }],
+            parserOptions: { ecmaVersion: 8 }
+        },
+        {
+            code: "async function a(){}",
+            options: [{ allow: ["asyncFunctions"] }],
+            parserOptions: { ecmaVersion: 8 }
```
Comment (yeonjuan) [inline tests/lib/rules/no-empty-function.js]: Could you move these test cases in the reduced `array` for consistency?
I think test cases in here are for the `non-empty` function cases. e.g.`var foo = () => 0;`

https://github.com/eslint/eslint/blob/f1525dc45dfdbbe31e724671270785b41cffc6bd/tests/lib/rules/no-empty-function.js#L90-L95

Comment (anikethsaha) [reply tests/lib/rules/no-empty-function.js]: what I understood from this test setup is that these cases are invalid cases and these will be converted into valid and invalid cases. right !

Comment (anikethsaha) [reply tests/lib/rules/no-empty-function.js]: > Would you mind updating the documentation also? - [no-empty-function](https://github.com/eslint/eslint/blob/master/docs/rules/no-empty-function.md#options)

sure !

**Comment by yeonjuan** [comment]
@anikethsaha 
Thanks for updating the documentation.
Maybe I didn't describe well at the [comment before](https://github.com/eslint/eslint/pull/13036#discussion_r392380739). sorry 😂



> what I understood from this test setup is that these cases are invalid cases and these will be converted into valid and invalid cases. right !

What I meant was that the newly added tests should be in the array at the third parameter of `ruleTester.run()`, not in the second parameter of  `Array.prototype.reduce()`.

Because the test cases at the second parameter of `Array.prototype.reduce()` are not converted into `invalid` and `valid` cases.
- the cases newly added are not passed in the second parameter(`item`) of `toValidInvalid(patterns, item)`


The current test setup seems to like this:
```js
ruleTester.run("no-empty-function", rule, [
   /* 
    * These elements are the test cases that need to be tested with each option and various invalid, valid form.
    * So these elements are converted into other forms by 'toValidInvalid()'
    */
    ,
]).reduce(toValidInvalid, {
  valid: [
    /*
     * These are always valid. So don't need to be converted  
     * That is the reason why "var foo = () => 0;" is here because it is not an empty function.
     */
    {
        code: "var foo = () => 0;", // it has return value '0'. always valid
        parserOptions: { ecmaVersion: 6 }
     }
   ] 

}));

```

**Comment by yeonjuan** [comment]
- [contributing guideline](https://eslint.org/docs/developer-guide/contributing/pull-requests#step-2-make-your-changes).
> . If the pull request addresses an issue, then the issue number should be mentioned at the end. 

And it would be nice if we mentioned the issue number in the PR title.

**Comment by anikethsaha** [comment]
@yeonjuan done 👍

File: tests/lib/rules/no-empty-function.js [hunk 2]
```diff
@@ -267,6 +269,16 @@ ruleTester.run("no-empty-function", rule, [
         {
             code: "var foo = () => 0;",
             parserOptions: { ecmaVersion: 6 }
+        },
+        {
+            code: "const foo = { async method() {} }",
```
Comment (kaicataldo) [inline tests/lib/rules/no-empty-function.js]: Can we add more tests for this change? It doesn't look like we're adding any new invalid tests right now.

File: docs/rules/no-empty-function.md [hunk 3]
```diff
@@ -177,6 +177,8 @@ This rule has an option to allow specific kinds of functions to be empty.
     * `"getters"` - Getters.
     * `"setters"` - Setters.
     * `"constructors"` - Class constructors.
+    * `"asyncFunctions"` -  Async Functions.
+    * `"asyncMethods"` - Class methods and method shorthands of object literals with async
```
Comment (kaicataldo) [inline docs/rules/no-empty-function.md]:
```suggestion
    * `"asyncMethods"` - Async class methods and method shorthands of object literals.
```

File: docs/rules/no-empty-function.md [hunk 4]
```diff
@@ -177,6 +177,8 @@ This rule has an option to allow specific kinds of functions to be empty.
     * `"getters"` - Getters.
     * `"setters"` - Setters.
     * `"constructors"` - Class constructors.
+    * `"asyncFunctions"` -  Async Functions.
```
Comment (kaicataldo) [inline docs/rules/no-empty-function.md]:
```suggestion
    * `"asyncFunctions"` - Async Functions.
```

**Comment by anikethsaha** [comment]
cc @kaicataldo @yeonjuan I added more tests and did the docs changes as request.

Do let me know if anything else needs to be changed!

File: docs/rules/no-empty-function.md [hunk 5]
```diff
@@ -307,6 +309,35 @@ class A {
 }
 ```
 
+#### allow: asyncFunctions
+
+Examples of **correct** code for  `{ "allow": ["asyncFunctions"] }` options:
```
Comment (yeonjuan) [inline docs/rules/no-empty-function.md]:
```suggestion
Examples of **correct** code for `{ "allow": ["asyncFunctions"] }` options:
```

File: docs/rules/no-empty-function.md [hunk 6]
```diff
@@ -307,6 +309,35 @@ class A {
...
+```js
+/*eslint no-empty-function: ["error", { "allow": ["asyncFunctions"] }]*/
+/*eslint-env es6*/
+
+async function a(){}
+```
+
+#### allow: asyncMethods
+
+Examples of **correct** code for  `{ "allow": ["asyncMethods"] }` options:
```
Comment (yeonjuan) [inline docs/rules/no-empty-function.md]:
```suggestion
Examples of **correct** code for `{ "allow": ["asyncMethods"] }` options:
```

File: docs/rules/no-empty-function.md [hunk 7]
```diff
@@ -177,6 +177,8 @@ This rule has an option to allow specific kinds of functions to be empty.
     * `"getters"` - Getters.
     * `"setters"` - Setters.
     * `"constructors"` - Class constructors.
+    * `"asyncFunctions"` - Async Functions.
```
Comment (yeonjuan) [inline docs/rules/no-empty-function.md]:
```suggestion
    * `"asyncFunctions"` - Async functions.
```

**Comment by anikethsaha** [comment]
cc @yeonjuan done

File: docs/rules/no-empty-function.md [hunk 8]
```diff
@@ -307,6 +309,35 @@ class A {
 }
 ```
 
+#### allow: asyncFunctions
+
+Examples of **correct** code for the `{ "allow": ["asyncFunctions"] }` options:
+
+```js
+/*eslint no-empty-function: ["error", { "allow": ["asyncFunctions"] }]*/
+/*eslint-env es6*/
```
[the remark this answers is not in the document]
Comment (anikethsaha) [reply docs/rules/no-empty-function.md]: ops ! yes es6 doesnt parse it . 

I will change it . thanks for pointing 👍

**Comment by kaicataldo** [comment]
Thanks for contributing to ESLint!

**Pull request merged**